### ISSUE TITLE
As a developer, I want an API method enable specifying of non-registered context products

### ISSUE THREAD

**Issue by fran-vallejo** [NASA-PDS/validate#241]
**Is your feature request related to a problem? Please describe.**
I'm frustrated with the way the registered / non-registered product files are passed to validate when validate is integrated as part of another application.

**Describe the solution you'd like**
To have a public method in ValidateLauncher which takes the URL to the registered / non registered product file as argument and includes them in the validation process. A method like: 

- `public void setContextProducts(URL registeredContextProductsFile, URL nonRegisteredContextProductsFile)` 

or a set of methods like:

- `public void addRegisteredContextProducts(URL registeredContextProductsFile)`
- `public void addNonRegisteredContextProducts(URL nonRegisteredContextProductsFile)`

**Additional context**
BepiColombo has an application to pack pds4 products (pds4-packager) and it uses validate as its main validation step. I have integrated validate using `gov.nasa.pds.validate.ValidateLauncher` as entry point as it provides public methods for almost the same features the command line exposes.
To correctly validate BepiColombo products I need validate to use pds' context products file (which they are available in validate.jar - util/registered_context_products.json) and esa not-yet-registered-in-pds context products file (which they are available in other class path location).

The main problem I have at this point is that `ValidateLauncher` does not provide a public method to load nonRegisteredContextProducts file so I have to abuse `ValidateLauncher` to have them loaded. 

Moreover, to load registered / non registered context products, the corresponding files should be available in the directory pointed by the `resources.home` property. For me it would be more convenient if `ValidateLauncher` would load them form a couple of URL locations (like it does with the catalog file).

I use the following code to set up validate:

```
public Stream<ValidationMessage> apply(Pds4Label label) {

		try {
			// Copy context product to java.io.tmpdir directory and set resource.home property as expected by validate
			
			final File registeredContextProductsFile = new File(System.getProperty("java.io.tmpdir") + File.separator + "registered_context_products.json");
			final File localContextProductsFile = new File(System.getProperty("java.io.tmpdir") + File.separator + "local_context_products.json");
			
			FileUtils.copyURLToFile(registeredContextProductsUrl, registeredContextProductsFile);
			FileUtils.copyURLToFile(localContextProductsUrl, localContextProductsFile);
			System.setProperty("resources.home", System.getProperty("java.io.tmpdir"));
			
			// Configure validate
			
			final ValidateLauncher launcher = new ValidateLauncher();
			
			launcher.setUpdateRegisteredProducts(true);
			launcher.setNonRegisteredProducts(true);
			
			launcher.setCatalogs(Arrays.asList(catalogUrl.toExternalForm()));
			launcher.setSkipProductValidation(!checkData);
			launcher.setForce(true);			
			launcher.setMaxErrors(Long.MAX_VALUE);
			launcher.setReportStyle("json");
			launcher.setTargets(Arrays.asList(label.getLabelLocation().toExternalForm()));
			launcher.setupReport();
			
			// Send report to a string writer to retrieve later
				
			final Report report = launcher.getReport();
			final StringWriter writer = new StringWriter();
				
			report.setOutput(writer);
			report.printHeader();
			
			// Work-around to register non registered products file
			
			final Field nonRegisteredProductsFileField = ValidateLauncher.class.getDeclaredField("nonRegisteredProductsFile");
			
			nonRegisteredProductsFileField.setAccessible(true);
			nonRegisteredProductsFileField.set(launcher, localContextProductsFile);
			
			// Work-around to avoid context validation to raise a NullPointerException 
			
			final Method setRegisteredProducts = ValidateLauncher.class.getDeclaredMethod("setRegisteredProducts");
			
			setRegisteredProducts.setAccessible(true);
			setRegisteredProducts.invoke(launcher);

			// Do validation
				
			launcher.doValidation(new HashMap<URL, String>());
				
			report.printFooter();
				
			writer.close();
									
			// Parse JSON reply
				
			final ValidationReport validationReport = buildValidationReport(writer);
			
			// Return validation messages
				
			return validationReport.getMessages().stream()
					.filter(this::filterNonRegisteredProductWarning)
					.filter(this::filterInvalidGroupLengthErrors);
		}
		catch(IOException e) {
			throw new UncheckedIOException(e);
		}
		catch(ReflectiveOperationException e) {
			throw new IllegalStateException(e);
		}
		catch(Exception e) {
			throw new RuntimeException(e);
		}
	}
```

The requested change would:

1. Remove the need to copy resources that are already available in the class path and can be loaded directly.
1. Remove the need to abuse `ValidateLauncher` code calling private methods and fields through reflection. 

I would be also happy to re-do the integration code if there is a recommended way to integrate validate.

---

## Applicable requirements
🦄 #308 

## Engineering Details
Will add getter/setter classes to the `ValidateLauncher` class.

**Note for I&T Team:** I&T will be completed by customer as they can easily drop this into their pipeline to verify it meets their requirement needs

**Comment by jordanpadams** [comment]
@fran-vallejo thanks so much for the feature request! I don't think we will be able to get this into our next feature release, but we will look to add this into our next build cycle.

also, apologies for the lack of usable API interfaces with Validate. with many developers writing the software, it looks like the API got away from us a bit. hopefully down the road we can get an intern in here to help us refactor some of these components back into the `Validate.java` class so we can have a proper API interface for users.

**Comment by jordanpadams** [comment]
@fran-vallejo the latest SNAPSHOT should contain this update. let me know how it checks out: https://github.com/NASA-PDS/validate/releases/tag/v2.2.4-SNAPSHOT

**Comment by fran-vallejo** [comment]
Hi @jordanpadams I have checked the latest snapshot and I would need a modification for our integration to work. 

From our integration code, calling doValidate() on gov.nasa.pds.validate.ValidateLauncher after calling setNonRegisteredProductsFile() and setRegisteredProductsFile() raises a NullPointerException. 

I have been debugging the code and I realised that on line 173  gov.nasa.pds.tools.validate.rule.pds4.ContextProductReferenceValidationRule.checkContextReferences() when checkContextReferences() retrieves the registered products from the context it gets a null object, causing a NullPointerException on line 220.

It seems that although the register / non registered context products have been set, they have not been parsed as the method used for parsing and put them in the context- ValidateLauncher.setRegisteredProducts() - is called in ValidateLauncher.processMain(String[] args) instead of in ValidateLaucher.doValidation() which is used for the integration.

I have tried to change the integration code to use ValidateLaucher.processMain(String args[]) instead of ValidateLauncher.doValidation() to work-around this error but that is not possible, mainly because processMain does it own report set-up and the integration code needs to customise the report before it writes anything.

To get the integration code working I would need that either ValidateLauncher.setRegisteredProducts() becomes public instead of private [so I can call it from the integration code before calling ValidateLauncher.doValidate()] or that the call to ValidateLauncher.setRegisteredProducts() in ValidateLauncher.processMain(String[] args) - line 1622 - gets moved inside ValidateLauncher.doValidate(). Could it be possible to include one of those modifications in the snapshot ?

Many thanks for your support.
